spec/geoserver_publisher_spec.rb

```diff
@@ -0,0 +1,139 @@
+      end
+    end
+
+    context 'ENV[GEOSERVER_*_URL] are set' do
+      before do
+        ENV['GEOSERVER_URL'] = 'https://public_user:public_pass@geoserver-public.lib.berkeley.edu/geoserver'
+        ENV['GEOSERVER_SECURE_URL'] = 'https://secure_user:secure_pass@geoserver-secure.lib.berkeley.edu/geoserver'
+      end
+
+      it 'falls back to GEOSERVER_URL' do
+        expect(subject.connection.config['url']).to eq 'https://geoserver-public.lib.berkeley.edu/geoserver/rest/'
+        expect(subject.connection.config['user']).to eq 'public_user'
+        expect(subject.connection.config['password']).to eq 'public_pass'
+      end
+
+      context 'it is secure' do
+        let(:default) { :geoserver_secure_url }
+
+        it 'falls back to GEOSERVER_SECURE_URL' do
+          expect(subject.connection.config['url']).to eq 'https://geoserver-secure.lib.berkeley.edu/geoserver/rest/'
+          expect(subject.connection.config['user']).to eq 'secure_user'
+          expect(subject.connection.config['password']).to eq 'secure_pass'
+        end
+      end
+    end
+
+    context 'nothing is given and ENV are not set' do
+      it 'falls back to the Config defaults' do
+        expect(subject.connection.config['url']).to eq 'http://geoserver:8080/geoserver/rest/'
+        expect(subject.connection.config['user']).to eq 'admin'
+        expect(subject.connection.config['password']).to eq 'geoserver'
+      end
+
+      context 'it is secure' do
+        let(:default) { :geoserver_secure_url }
+
+        it 'falls back to the Config defaults' do
+          expect(subject.connection.config['url']).to eq 'http://geoserver-secure:8080/geoserver/rest/'
+          expect(subject.connection.config['user']).to eq 'admin'
+          expect(subject.connection.config['password']).to eq 'geoserver'
+        end
+      end
+    end
+  end
+
+  describe '.parse_connection_string' do
+    it 'correctly infers the REST URL from a baseurl' do
+      connstring = 'http://admin:geoserver@geoserver:8080/geoserver'
+      url, user, password = described_class.parse_connection_string(connstring)
+      expect(url).to eq 'http://geoserver:8080/geoserver/rest/'
+      expect(user).to eq 'admin'
+      expect(password).to eq 'geoserver'
+    end
+  end
+
+  describe '#create_workspace' do
+    let(:workspace_name) { "test-UCB-#{Random.rand(1000)}" }
+    let(:workspace_client) { Geoserver::Publish::Workspace.new(subject.connection) }
+
+    after { workspace_client.delete(workspace_name:) }
+
+    it 'creates a workspace if one does not already exist' do
+      expect { subject.create_workspace }
+        .to change { workspace_client.find(workspace_name:) }
+        .from(nil)
+    end
+
+    it 'is a no-op if the workspace already exists' do
+      workspace_client.create(workspace_name:)
+      expect { subject.create_workspace }
+        .not_to change { workspace_client.find(workspace_name:) }
+    end
+  end
+
+  describe 'publish' do
```

Comment:
These tests all hit the actual running GeoServers, so they're pretty valuable. We need a few more end-to-end tests like these (especially for the DataHandler and the logic in the Cli.all method).